File: app/src/main/java/com/robifr/ledger/ui/searchcustomer/SearchCustomerFragment.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.searchcustomer;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.activity.OnBackPressedCallback;
import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentFactory;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import com.robifr.ledger.R;
import com.robifr.ledger.databinding.SearchableFragmentBinding;
import com.robifr.ledger.ui.BackStack;
import com.robifr.ledger.ui.FragmentResultKey;
import com.robifr.ledger.ui.searchcustomer.recycler.SearchCustomerAdapter;
import com.robifr.ledger.ui.searchcustomer.viewmodel.SearchCustomerViewModel;
import com.robifr.ledger.util.Compats;
import java.util.Objects;

public class SearchCustomerFragment extends Fragment implements SearchView.OnQueryTextListener {
  public enum Request implements FragmentResultKey {
    SELECT_CUSTOMER;

    @Override
    @NonNull
    public String key() {
      return FragmentResultKey.generateKey(this);
    }
  }

  public enum Result implements FragmentResultKey {
    SELECTED_CUSTOMER_ID;

    @Override
    @NonNull
    public String key() {
      return FragmentResultKey.generateKey(this);
    }
  }

  @NonNull private final OnBackPressedHandler _onBackPressed = new OnBackPressedHandler();
  @Nullable private final String _initialQuery;
  @Nullable private SearchableFragmentBinding _fragmentBinding;
  @Nullable private SearchCustomerAdapter _adapter;
  @ColorInt private int _normalStatusBarColor;

  @Nullable private SearchCustomerViewModel _searchCustomerViewModel;
  @Nullable private SearchCustomerViewModelHandler _viewModelHandler;

  /** Default constructor when configuration changes. */
  public SearchCustomerFragment() {
    this(null);
  }

  private SearchCustomerFragment(@Nullable String initialQuery) {
    this._initialQuery = initialQuery;
  }

  @Override
  public View onCreateView(
      @NonNull LayoutInflater inflater,
      @Nullable ViewGroup container,
      @Nullable Bundle savedInstance) {
    Objects.requireNonNull(inflater);

    this._fragmentBinding = SearchableFragmentBinding.inflate(inflater, container, false);
    return this._fragmentBinding.getRoot();
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstance) {
    Objects.requireNonNull(view);
    Objects.requireNonNull(this._fragmentBinding);

    this._adapter = new SearchCustomerAdapter(this);
    this._normalStatusBarColor = this.requireActivity().getWindow().getStatusBarColor();
    this._searchCustomerViewModel =
        new ViewModelProvider(this, new SearchCustomerViewModel.Factory(this.requireContext()))
            .get(SearchCustomerViewModel.class);
    this._viewModelHandler =
        new SearchCustomerViewModelHandler(this, this._searchCustomerViewModel);

    this.requireActivity()
        .getOnBackPressedDispatcher()
        .addCallback(this.getViewLifecycleOwner(), this._onBackPressed);
    this.requireActivity()
        .getWindow() // Match status bar color with toolbar.
        .setStatusBarColor(this.requireContext().getColor(R.color.surface));
    this._fragmentBinding.toolbar.setNavigationOnClickListener(
        v -> this._onBackPressed.handleOnBackPressed());
    this._fragmentBinding.seachView.setQueryHint(this.getString(R.string.text_search_customers));
    this._fragmentBinding.seachView.setOnQueryTextListener(this);
    this._fragmentBinding.seachView.requestFocus();
    this._fragmentBinding.noResultsImage.image.setImageResource(R.drawable.image_noresultsfound);
    this._fragmentBinding.noResultsImage.title.setText(R.string.text_no_results_found);
    this._fragmentBinding.noResultsImage.description.setText(
        this.getString(R.string.text_cant_find_any_matching_customers));
    this._fragmentBinding.recyclerView.setLayoutManager(
        new LinearLayoutManager(this.requireContext()));
    this._fragmentBinding.recyclerView.setAdapter(this._adapter);
    this._fragmentBinding.recyclerView.setItemViewCacheSize(0);

    if (this._initialQuery != null) {
      this._fragmentBinding.seachView.setQuery(this._initialQuery, true);
    } else {
      Compats.showKeyboard(this.requireContext(), this._fragmentBinding.seachView);
    }
  }

  @Override
  public boolean onQueryTextSubmit(@NonNull String query) {
    return false;
  }

  @Override
  public boolean onQueryTextChange(@NonNull String newText) {
    Objects.requireNonNull(this._searchCustomerViewModel);

    this._searchCustomerViewModel.onSearch(newText);
    return true;
  }

  @NonNull
  public SearchableFragmentBinding fragmentBinding() {
    return Objects.requireNonNull(this._fragmentBinding);
  }

  @NonNull
  public SearchCustomerAdapter adapter() {
    return Objects.requireNonNull(this._adapter);
  }

  @NonNull
  public SearchCustomerViewModel searchCustomerViewModel() {
    return Objects.requireNonNull(this._searchCustomerViewModel);
  }

  public void finish() {
    Objects.requireNonNull(this._fragmentBinding);

    if (this.requireActivity() instanceof BackStack navigation
        && navigation.currentTabStackTag() != null) {
      Compats.hideKeyboard(this.requireContext(), this.requireView().findFocus());
      this.requireActivity().getWindow().setStatusBarColor(this._normalStatusBarColor);
      navigation.popFragmentStack(navigation.currentTabStackTag());
    }
  }

  public static class Factory extends FragmentFactory {
    @Nullable private final String _initialQuery;

    public Factory(@Nullable String initialQuery) {
      this._initialQuery = initialQuery;
    }

    @Override
    @NonNull
    public Fragment instantiate(@NonNull ClassLoader classLoader, @NonNull String className) {
      Objects.requireNonNull(classLoader);
      Objects.requireNonNull(className);

      return (className.equals(SearchCustomerFragment.class.getName()))
          ? new SearchCustomerFragment(this._initialQuery)
          : super.instantiate(classLoader, className);
    }
  }

  private class OnBackPressedHandler extends OnBackPressedCallback {
    public OnBackPressedHandler() {
      super(true);
    }

    @Override
    public void handleOnBackPressed() {
      SearchCustomerFragment.this.searchCustomerViewModel().onCustomerSelected(null);
    }
  }
}

File: app/src/main/java/com/robifr/ledger/ui/searchcustomer/SearchCustomerFragment.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.searchcustomer

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.widget.SearchView
import androidx.core.graphics.Insets
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.isVisible
import androidx.core.view.updatePadding
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.snackbar.Snackbar
import com.robifr.ledger.R
import com.robifr.ledger.databinding.SearchableFragmentBinding
import com.robifr.ledger.ui.FragmentResultKey
import com.robifr.ledger.ui.RecyclerAdapterState
import com.robifr.ledger.ui.SnackbarState
import com.robifr.ledger.ui.customer.CustomerMenu
import com.robifr.ledger.ui.search.viewmodel.SearchState
import com.robifr.ledger.ui.search.viewmodel.SearchViewModel
import com.robifr.ledger.ui.searchcustomer.recycler.SearchCustomerAdapter
import com.robifr.ledger.ui.searchcustomer.viewmodel.SearchCustomerResultState
import com.robifr.ledger.ui.searchcustomer.viewmodel.SearchCustomerState
import com.robifr.ledger.ui.searchcustomer.viewmodel.SearchCustomerViewModel
import com.robifr.ledger.util.hideKeyboard
import com.robifr.ledger.util.showKeyboard
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class SearchCustomerFragment : Fragment(), SearchView.OnQueryTextListener {
  private var _fragmentBinding: SearchableFragmentBinding? = null
  val fragmentBinding: SearchableFragmentBinding
    get() = _fragmentBinding!!

  val searchCustomerViewModel: SearchCustomerViewModel by viewModels()
  private val _searchViewModel: SearchViewModel by
      viewModels({ if (parentFragment !is NavHostFragment) requireParentFragment() else this })
  private lateinit var _customerMenu: CustomerMenu
  private lateinit var _adapter: SearchCustomerAdapter
  private lateinit var _onBackPressed: OnBackPressedHandler

  override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?
  ): View {
    _fragmentBinding = SearchableFragmentBinding.inflate(inflater, container, false)
    _customerMenu = CustomerMenu(this, searchCustomerViewModel::onCustomerMenuDialogClosed)
    _adapter = SearchCustomerAdapter(this)
    _onBackPressed = OnBackPressedHandler(this)
    return fragmentBinding.root
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    ViewCompat.setOnApplyWindowInsetsListener(fragmentBinding.root) { _, insets ->
      val systemBarInsets: Insets = insets.getInsets(WindowInsetsCompat.Type.systemBars())
      val windowInsets: Insets =
          insets.getInsets(
              WindowInsetsCompat.Type.displayCutout() or WindowInsetsCompat.Type.navigationBars())
      fragmentBinding.appBarLayout.updatePadding(
          top = systemBarInsets.top, left = windowInsets.left, right = windowInsets.right)
      fragmentBinding.noResultsImageContainer.updatePadding(
          left = windowInsets.left, right = windowInsets.right)
      fragmentBinding.recyclerView.updatePadding(
          left = windowInsets.left, right = windowInsets.right)
      WindowInsetsCompat.CONSUMED
    }
    requireActivity().onBackPressedDispatcher.addCallback(viewLifecycleOwner, _onBackPressed)
    fragmentBinding.appBarLayout.isVisible =
        searchCustomerViewModel.uiState.safeValue.isToolbarVisible
    fragmentBinding.toolbar.setNavigationOnClickListener { _onBackPressed.handleOnBackPressed() }
    fragmentBinding.searchView.queryHint = getString(R.string.searchCustomer)
    fragmentBinding.searchView.setOnQueryTextListener(this)
    fragmentBinding.noResultsImage.image.setImageResource(R.drawable.image_search_3d)
    fragmentBinding.noResultsImage.title.setText(R.string.searchCustomer_noResultsFound)
    fragmentBinding.noResultsImage.description.setText(
        R.string.searchCustomer_noResultsFound_description)
    fragmentBinding.recyclerView.layoutManager = LinearLayoutManager(requireContext())
    fragmentBinding.recyclerView.adapter = _adapter
    fragmentBinding.recyclerView.setItemViewCacheSize(0)
    searchCustomerViewModel.snackbarState.observe(viewLifecycleOwner, ::_onSnackbarState)
    searchCustomerViewModel.resultState.observe(viewLifecycleOwner, ::_onResultState)
    searchCustomerViewModel.uiState.observe(viewLifecycleOwner, ::_onUiState)
    searchCustomerViewModel.recyclerAdapterState.observe(
        viewLifecycleOwner, ::_onRecyclerAdapterState)
    _searchViewModel.uiState.observe(viewLifecycleOwner, ::_onSearchUiState)

    if (searchCustomerViewModel.uiState.safeValue.isToolbarVisible) {
      if (searchCustomerViewModel.uiState.safeValue.initialQuery.isNotEmpty()) {
        fragmentBinding.searchView.setQuery(
            searchCustomerViewModel.uiState.safeValue.initialQuery, true)
      } else {
        fragmentBinding.searchView.requestFocus()
        fragmentBinding.searchView.showKeyboard()
      }
    }
  }

  override fun onQueryTextSubmit(query: String?): Boolean = false

  override fun onQueryTextChange(newText: String?): Boolean {
    newText?.let { searchCustomerViewModel.onSearch(it) }
    return true
  }

  fun finish() {
    requireView().hideKeyboard()
    findNavController().popBackStack()
  }

  private fun _onSnackbarState(state: SnackbarState) {
    Snackbar.make(
            fragmentBinding.root as View,
            state.messageRes.toStringValue(requireContext()),
            Snackbar.LENGTH_LONG)
        .show()
  }

  private fun _onResultState(state: SearchCustomerResultState) {
    if (searchCustomerViewModel.uiState.safeValue.isSelectionEnabled) {
      parentFragmentManager.setFragmentResult(
          Request.SELECT_CUSTOMER.key(),
          Bundle().apply {
            state.selectedCustomerId?.let { putLong(Result.SELECTED_CUSTOMER_ID_LONG.key(), it) }
          })
    }
    finish()
  }

  private fun _onUiState(state: SearchCustomerState) {
    // FIXME: Customer menu dialog doesn't get retained (it close) when configuration changes.
    if (state.isCustomerMenuDialogShown) {
      state.selectedCustomerMenu?.let {
        _customerMenu.showDialog(it, searchCustomerViewModel::onDeleteCustomer)
      }
    } else {
      _customerMenu.dismissDialog()
    }
    fragmentBinding.noResultsImageContainer.isVisible = state.isNoResultFoundIllustrationVisible
    fragmentBinding.recyclerView.isVisible = state.isRecyclerViewVisible
  }

  private fun _onRecyclerAdapterState(state: RecyclerAdapterState) {
    when (state) {
      is RecyclerAdapterState.DataSetChanged -> _adapter.notifyDataSetChanged()
      is RecyclerAdapterState.ItemChanged ->
          state.indexes.forEach { _adapter.notifyItemChanged(it) }
    }
  }

  private fun _onSearchUiState(state: SearchState) {
    searchCustomerViewModel.onSearchUiStateChanged(state)
  }

  enum class Arguments : FragmentResultKey {
    IS_SELECTION_ENABLED_BOOLEAN,
    IS_TOOLBAR_VISIBLE_BOOLEAN,
    INITIAL_QUERY_STRING,
    INITIAL_SELECTED_CUSTOMER_IDS_LONG_ARRAY
  }

  enum class Request : FragmentResultKey {
    SELECT_CUSTOMER
  }

  enum class Result : FragmentResultKey {
    SELECTED_CUSTOMER_ID_LONG
  }
}

private class OnBackPressedHandler(private val _fragment: SearchCustomerFragment) :
    OnBackPressedCallback(true) {
  override fun handleOnBackPressed() {
    _fragment.searchCustomerViewModel.onCustomerSelected(null)
  }
}
